OC-4074 Update User

hosh opened the pull request:
Wiring for updating users

hosh committed:
[OC-4074] Added update_from_ejson() and set_updated() for #chef_user

seth commented on src/chef_object.erl:
```diff
@@ -315,8 +315,18 @@ update_from_ejson(#chef_cookbook_version{org_id = OrgId,
-
+    Name = ej:get({<<"name">>}, UserData),
+    IsAdmin = ej:get({<<"admin">>}, UserData) =:= true,
+    Email = ej:get({<<"email">>}, UserData),
```
The "email" key may not always be present in the EJSON. I think we need to have a defualt value of the atom 'null' for the right thing to happen if it is missing.

hosh replied:
As discussed, "email" is not in the chef_db query and we're not using it right now, taking this out.

seth commented on src/chef_object.erl:
```diff
@@ -315,8 +315,18 @@ update_from_ejson(#chef_cookbook_version{org_id = OrgId,
+    Name = ej:get({<<"name">>}, UserData),
+    IsAdmin = ej:get({<<"admin">>}, UserData) =:= true,
+    Email = ej:get({<<"email">>}, UserData),
+    %% IsValidator = ej:get({<<"validator">>}, UserData) =:= true,
```
Why is this commented out? I think we want to be able to update these fields. Is the reason because we do the update to the incoming user record in the resource and so don't need to pull the data out of the EJSON? If that's the case, please remove the commented code and add a comment describing how this stuff currently works.

hosh replied:
As discussed, this was left-over from copy-and-paste from client. I'll remove it.

seth commented:
Also, I think we have an extra chef_user record clause in chef_objects at line 417 in set_updated.

seth commented on src/chef_object.erl:
```diff
@@ -403,7 +413,10 @@ set_updated(#chef_user{} = Object, ActorId) ->
     Now = sql_date(now),
-    Object#chef_cookbook_version{updated_at = Now, last_updated_by = ActorId}.
+    Object#chef_cookbook_version{updated_at = Now, last_updated_by = ActorId};
+set_updated(#chef_user{} = Object, ActorId) ->
```
There is already a fun head for chef_user above :)

Commit:
Fix spec and extra fun head for set_updated

hosh committed:
[OC-4074] Cleaned up comments and left-overs

hosh committed:
[OC-4074] Refactored password and key handling for users

seth commented on src/chef_user.erl:
```diff
@@ -132,5 +134,13 @@ key_version(<<"-----BEGIN RSA PUBLIC KEY", _Bin/binary>>) ->
     ?KEY_VERSION.
 
+password_data(#chef_user{hashed_password = HashedPassword,
+                                         salt = Salt,
```
Please adjust formatting here, should align with column after`{`.

seth commented:
Aside from the minor formatting fix, this looks good and local chef_objects tests pass for me. After you fix the whitespace/alignment, merge when ready.

hosh committed:
[OC-4075] Fixed spec for updating user

- Broke out chef_object:update_from_ejson() to
  chef_user:update_from_ejson() in order to avoid nasty spec calls
- chef_user:update_from_ejson() will be directly called from
  chef_wm_named_user:update_from_json
- This is temporary until we can properly break up
  chef_wm_base:update_from_json into composable functions
- Fixed spec to reflect reality

hosh committed:
[OC-4074] Updated spec for chef_user:update_from_ejson()

seth commented:
merge it :)
